compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinOptimizationLowering.kt

```diff
@@ -43,56 +45,72 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
-                    if (hasNoSideEffectsForNullCompare(nonNullArgument)) {
-                        constFalse
-                    } else {
-                        IrBlockImpl(expression.startOffset, expression.endOffset, expression.type, expression.origin).apply {
-                            statements.add(nonNullArgument.coerceToUnitIfNeeded(nonNullArgument.type.toKotlinType(), context.irBuiltIns))
-                            statements.add(constFalse)
+
+                if (isNegation(expression, context) && isNegation(expression.dispatchReceiver!!, context)) {
+                    return (expression.dispatchReceiver as IrCall).dispatchReceiver!!
+                }
+
+                getOperandsIfCallToEqeqOrEquals(expression)?.let {
+                    val left = it.first
+                    val right = it.second
+
+                    return when {
+                        left.isNullConst() && right.isNullConst() ->
+                            IrConstImpl.constTrue(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
+
+                        it.second.isNullConst() && it.first.type.unboxInlineClass().isPrimitiveType() ||
+                                it.first.isNullConst() && it.second.type.unboxInlineClass().isPrimitiveType() -> {
+                            val nonNullArgument = if (left.isNullConst()) right else left
+                            val constFalse =
+                                IrConstImpl.constFalse(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
+                            if (hasNoSideEffectsForNullCompare(nonNullArgument)) {
+                                constFalse
+                            } else {
+                                IrBlockImpl(expression.startOffset, expression.endOffset, expression.type, expression.origin).apply {
+                                    statements.add(
+                                        nonNullArgument.coerceToUnitIfNeeded(
+                                            nonNullArgument.type.toKotlinType(),
+                                            context.irBuiltIns
+                                        )
+                                    )
+                                    statements.add(constFalse)
+                                }
+                            }
+
                         }
+
+                        else -> expression
                     }
-                } else if (isNullCheckOfConstant(expression, context)) {
-                    if (expression.getValueArgument(0)!!.isNullConst() && expression.getValueArgument(1)!!.isNullConst()) {
-                        IrConstImpl.constTrue(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
-                    } else {
-                        IrConstImpl.constFalse(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
```

Comment:
false optimization is lost

Reply:
I believe the `isNullCheckOfConstant` check is subsumed by the call to `hasNoSideEffectsForNullCompare`. I.e. paths that hit line 89 (`null == null`) end at 82. Paths ending at 91 (`null == "constant"`, `4.equals(null)`, ...) now end at 90.